### DISCUSSION TITLE
[4.1] problems with code format in tinymce

### DISCUSSION THREAD

**Discussion by N6REJ** [joomla/joomla-cms#37863] [Q&A] [answered]
I'm having an issue where when I paste in formatted code in tinymce ( editor off ) and then look at it after saving all the formatting is gone.  Turning off "joomla filters" doesn't change the behavior.
```
<hr id = "system-readmore" />
<div id = "listing">
	<!-- header -->
	<div class = "row">
		<div class = "col gallery">
			<!-- single pic -->
			{field 1}
			<br />
			<!-- gallery -->
			{field 3}
		</div>
		<div class = "col-sm-4 side-specs">
			<div class = "card">
				<div class = "card-header">
					<h3 class = "price">{field 10}</h3>
					<h6 class = "text-muted" style = "text-align: right;">*Pricing excludes licensing and tax</h6>
				</div>
				<div class = "card-body">
					<ul>
						<li class = "vin">{field 14}</li>
						<li class = "mileage">{field 11}</li>
						<li class = "condition">{field 13}</li>
						<li class = "style">{field 15}</li>
						<li class = "engine">{field 16}</li>
						<li class = "transmission">{field 17}</li>
						<li class = "drivetrain">{field 18}</li>
						<li class = "doors">{field 19}</li>
						<li class = "fuel">{field 20}</li>
					</ul>
					<br />
					<div class = "description">{field 12}</div>
				</div>
			</div>
			<a href="#" class="btn btn-dark" style="width: 100%;">Call me</a>
		</div>
	</div>
	<!-- details -->
	<div class = "row leading">
		<div class = "col">
			<h1 class = "h1-listing title">{field 2}</h1>
			<h2 class = "h2-listing">
				<!-- price / mileage -->
				{field 10} {field 11}
			</h2>
			<br />
			<div class = "description">{field 12}</div>
		</div>
	</div>
	<!-- columnar content -->
	<div class = "row columnar">
		<div class = "col left">{left content}</div>
		<div class = "col right">{right content}</div>
	</div>
</div>
```
becomes 
![image](https://user-images.githubusercontent.com/1850089/169725316-dd68ffe0-526b-400a-9e52-129098a98744.png)

**Comment by brianteeman** [comment, marked as answer]
This is the intended and expectd behaviour of tinymce.

> Preserving white space in a rich text editor will never work. Just think about it, it would need to have some form of AI to figure out how the current document is formatted with white space and depending on that format apply the same indentation to new contents of the same type or it will be completely random behavior existing contents would be indented for a while until you modify it.
> 
> So it's better to just normalize the whitespace/formatting so that it will always be produced in the same way. We do that by removing all white space before after block elements and some whitespace that is redundant then re-apply line feeds before/after block elements. We don't indent with tabs/spaces since it would produce way to much HTML code for storage. We might add some plugin for that in the future that does the extra indentation.
> 
> Best regards,
> Spocke - Main developer of TinyMCE
> 
>

**Comment by N6REJ** [comment]
thanks to spocke for clearing this up.
